File: togo/internal/services/document_symbols.go
package services

// Builds an LSP DocumentSymbol tree (outline) from a parsed SourceFile. Kept
// separate from the server transport so it is unit-testable.
// Port of src/services/documentSymbols.ts.

import (
	"github.com/nikeee/cappu/internal/compiler"
	"github.com/nikeee/cappu/internal/lsp"
)

func dsRange(text string, lineStarts []int, pos, end int) lsp.Range {
	s := compiler.GetLineAndCharacterOfPosition(text, lineStarts, pos)
	e := compiler.GetLineAndCharacterOfPosition(text, lineStarts, end)
	return lsp.Range{
		Start: lsp.Position{Line: s.Line, Character: s.Character},
		End:   lsp.Position{Line: e.Line, Character: e.Character},
	}
}

func dsSymbol(name string, kind lsp.SymbolKind, node, selection *compiler.Node, text string, lineStarts []int, children []lsp.DocumentSymbol) lsp.DocumentSymbol {
	if name == "" {
		name = "<anonymous>"
	}
	// `node` is the declaration for every caller except fields (where it is the
	// VariableDeclarator and @Deprecated sits on the enclosing FieldDeclaration -
	// that branch tags the symbol itself, since parent pointers need binding).
	var tags []int
	if _, ok := compiler.ReadDeprecation(node); ok {
		tags = []int{lsp.SymbolTagDeprecated}
	}
	return lsp.DocumentSymbol{
		Name:           name,
		Kind:           kind,
		Range:          dsRange(text, lineStarts, node.Pos, node.End),
		SelectionRange: dsRange(text, lineStarts, selection.Pos, selection.End),
		Children:       children,
		Tags:           tags,
	}
}

func membersOf(members []*compiler.Node, text string, lineStarts []int) []lsp.DocumentSymbol {
	var result []lsp.DocumentSymbol
	for _, member := range members {
		result = append(result, memberSymbols(member, text, lineStarts)...)
	}
	return result
}

func memberSymbols(node *compiler.Node, text string, lineStarts []int) []lsp.DocumentSymbol {
	switch node.Kind {
	case compiler.ClassDeclaration, compiler.InterfaceDeclaration, compiler.EnumDeclaration,
		compiler.AnnotationTypeDeclaration, compiler.RecordDeclaration:
		return []lsp.DocumentSymbol{typeSymbol(node, text, lineStarts)}
	case compiler.MethodDeclaration:
		name := node.AsMethodDeclaration().Name
		return []lsp.DocumentSymbol{dsSymbol(name.AsIdentifier().Text, lsp.SymbolKindMethod, node, name, text, lineStarts, nil)}
	case compiler.ConstructorDeclaration:
		name := node.AsConstructorDeclaration().Name
		return []lsp.DocumentSymbol{dsSymbol(name.AsIdentifier().Text, lsp.SymbolKindConstructor, node, name, text, lineStarts, nil)}
	case compiler.CompactConstructorDeclaration:
		name := node.AsCompactConstructorDeclaration().Name
		return []lsp.DocumentSymbol{dsSymbol(name.AsIdentifier().Text, lsp.SymbolKindConstructor, node, name, text, lineStarts, nil)}
	case compiler.FieldDeclaration:
		// @Deprecated sits on the FieldDeclaration, not the per-name declarators.
		_, deprecated := compiler.ReadDeprecation(node)
		var out []lsp.DocumentSymbol
		for _, d := range node.AsFieldDeclaration().Declarators.Nodes {
			name := d.AsVariableDeclarator().Name
			s := dsSymbol(name.AsIdentifier().Text, lsp.SymbolKindField, d, name, text, lineStarts, nil)
			if deprecated {
				s.Tags = []int{lsp.SymbolTagDeprecated}
			}
			out = append(out, s)
		}
		return out
	case compiler.EnumConstantDeclaration:
		name := node.AsEnumConstantDeclaration().Name
		return []lsp.DocumentSymbol{dsSymbol(name.AsIdentifier().Text, lsp.SymbolKindEnumMember, node, name, text, lineStarts, nil)}
	case compiler.RecordComponent:
		name := node.AsRecordComponent().Name
		return []lsp.DocumentSymbol{dsSymbol(name.AsIdentifier().Text, lsp.SymbolKindField, node, name, text, lineStarts, nil)}
	default:
		return nil
	}
}

func typeSymbol(node *compiler.Node, text string, lineStarts []int) lsp.DocumentSymbol {
	var children []lsp.DocumentSymbol
	kind := lsp.SymbolKindClass
	var name *compiler.Node

	switch node.Kind {
	case compiler.InterfaceDeclaration:
		kind = lsp.SymbolKindInterface
		d := node.AsInterfaceDeclaration()
		name = d.Name
		children = append(children, membersOf(nodesOf(d.Members), text, lineStarts)...)
	case compiler.AnnotationTypeDeclaration:
		kind = lsp.SymbolKindInterface
		d := node.AsAnnotationTypeDeclaration()
		name = d.Name
		children = append(children, membersOf(nodesOf(d.Members), text, lineStarts)...)
	case compiler.EnumDeclaration:
		kind = lsp.SymbolKindEnum
		d := node.AsEnumDeclaration()
		name = d.Name
		children = append(children, membersOf(nodesOf(d.EnumConstants), text, lineStarts)...)
		children = append(children, membersOf(nodesOf(d.Members), text, lineStarts)...)
	case compiler.RecordDeclaration:
		d := node.AsRecordDeclaration()
		name = d.Name
		children = append(children, membersOf(nodesOf(d.RecordComponents), text, lineStarts)...)
		children = append(children, membersOf(nodesOf(d.Members), text, lineStarts)...)
	default:
		d := node.AsClassDeclaration()
		name = d.Name
		children = append(children, membersOf(nodesOf(d.Members), text, lineStarts)...)
	}

	return dsSymbol(name.AsIdentifier().Text, kind, node, name, text, lineStarts, children)
}

// GetDocumentSymbols builds the top-level outline for a source file.
func GetDocumentSymbols(sourceFile *compiler.Node, lineStarts []int) []lsp.DocumentSymbol {
	text := sourceFile.AsSourceFile().Text
	var result []lsp.DocumentSymbol
	for _, statement := range sourceFile.AsSourceFile().Statements.Nodes {
		switch statement.Kind {
		case compiler.ClassDeclaration, compiler.InterfaceDeclaration, compiler.EnumDeclaration,
			compiler.AnnotationTypeDeclaration, compiler.RecordDeclaration:
			result = append(result, typeSymbol(statement, text, lineStarts))
		}
	}
	return result
}

File: src/services/documentSymbols.ts
// Builds an LSP DocumentSymbol tree (outline) from a parsed SourceFile. Kept
// separate from the server transport so it is unit-testable.

import { type DocumentSymbol, type Range, SymbolKind, SymbolTag } from "vscode-languageserver-types";

import { getLineAndCharacterOfPosition } from "../compiler/lineMap.ts";
import { readDeprecation } from "../compiler/deprecation.ts";
import {
  type ClassDeclaration,
  type EnumConstantDeclaration,
  type EnumDeclaration,
  type FieldDeclaration,
  type Identifier,
  type InterfaceDeclaration,
  type MethodDeclaration,
  type Node,
  type NodeArray,
  type RecordComponent,
  type RecordDeclaration,
  type SourceFile,
  SyntaxKind,
} from "../compiler/types.ts";

function range(lineStarts: readonly number[], pos: number, end: number): Range {
  return {
    start: getLineAndCharacterOfPosition(lineStarts, pos),
    end: getLineAndCharacterOfPosition(lineStarts, end),
  };
}

function symbol(
  name: string,
  kind: SymbolKind,
  node: Node,
  selection: Node,
  lineStarts: readonly number[],
  children?: DocumentSymbol[],
): DocumentSymbol {
  // `node` is the declaration for every caller except fields (where it is the
  // VariableDeclarator and @Deprecated sits on the enclosing FieldDeclaration -
  // that branch tags the symbol itself, since parent pointers need binding).
  const deprecated = readDeprecation(node) !== undefined;
  return {
    name: name || "<anonymous>",
    kind,
    range: range(lineStarts, node.pos, node.end),
    selectionRange: range(lineStarts, selection.pos, selection.end),
    children,
    ...(deprecated ? { tags: [SymbolTag.Deprecated] } : {}),
  };
}

function nameText(node: { name?: Identifier }): string {
  return node.name?.text ?? "";
}

function membersOf(members: NodeArray<Node>, lineStarts: readonly number[]): DocumentSymbol[] {
  const result: DocumentSymbol[] = [];
  for (const member of members) {
    result.push(...memberSymbols(member, lineStarts));
  }
  return result;
}

function memberSymbols(node: Node, lineStarts: readonly number[]): DocumentSymbol[] {
  switch (node.kind) {
    case SyntaxKind.ClassDeclaration:
    case SyntaxKind.InterfaceDeclaration:
    case SyntaxKind.EnumDeclaration:
    case SyntaxKind.AnnotationTypeDeclaration:
    case SyntaxKind.RecordDeclaration:
      return [typeSymbol(node, lineStarts)];

    case SyntaxKind.MethodDeclaration: {
      const m = node as MethodDeclaration;
      return [symbol(nameText(m), SymbolKind.Method, node, m.name, lineStarts)];
    }
    case SyntaxKind.ConstructorDeclaration:
    case SyntaxKind.CompactConstructorDeclaration: {
      const c = node as unknown as { name: Identifier };
      return [symbol(c.name.text, SymbolKind.Constructor, node, c.name, lineStarts)];
    }
    case SyntaxKind.FieldDeclaration: {
      const f = node as FieldDeclaration;
      // @Deprecated sits on the FieldDeclaration, not the per-name declarators.
      const deprecated = readDeprecation(f) !== undefined;
      return f.declarators.map(d => {
        const s = symbol(d.name.text, SymbolKind.Field, d, d.name, lineStarts);
        if (deprecated) s.tags = [SymbolTag.Deprecated];
        return s;
      });
    }
    case SyntaxKind.EnumConstantDeclaration: {
      const e = node as EnumConstantDeclaration;
      return [symbol(e.name.text, SymbolKind.EnumMember, node, e.name, lineStarts)];
    }
    case SyntaxKind.RecordComponent: {
      const r = node as RecordComponent;
      return [symbol(r.name.text, SymbolKind.Field, node, r.name, lineStarts)];
    }
    default:
      return [];
  }
}

function typeSymbol(node: Node, lineStarts: readonly number[]): DocumentSymbol {
  const children: DocumentSymbol[] = [];
  let kind: SymbolKind = SymbolKind.Class;
  let name: Identifier;

  switch (node.kind) {
    case SyntaxKind.InterfaceDeclaration:
    case SyntaxKind.AnnotationTypeDeclaration:
      kind = SymbolKind.Interface;
      name = (node as InterfaceDeclaration).name;
      children.push(...membersOf((node as InterfaceDeclaration).members, lineStarts));
      break;
    case SyntaxKind.EnumDeclaration: {
      kind = SymbolKind.Enum;
      const e = node as EnumDeclaration;
      name = e.name;
      children.push(...membersOf(e.enumConstants, lineStarts));
      children.push(...membersOf(e.members, lineStarts));
      break;
    }
    case SyntaxKind.RecordDeclaration: {
      const r = node as RecordDeclaration;
      name = r.name;
      children.push(...membersOf(r.recordComponents, lineStarts));
      children.push(...membersOf(r.members, lineStarts));
      break;
    }
    default: {
      const c = node as ClassDeclaration;
      name = c.name;
      children.push(...membersOf(c.members, lineStarts));
      break;
    }
  }

  return symbol(name.text, kind, node, name, lineStarts, children);
}

/** Top-level outline for a source file. */
export function getDocumentSymbols(
  sourceFile: SourceFile,
  lineStarts: readonly number[],
): DocumentSymbol[] {
  const result: DocumentSymbol[] = [];
  for (const statement of sourceFile.statements) {
    switch (statement.kind) {
      case SyntaxKind.ClassDeclaration:
      case SyntaxKind.InterfaceDeclaration:
      case SyntaxKind.EnumDeclaration:
      case SyntaxKind.AnnotationTypeDeclaration:
      case SyntaxKind.RecordDeclaration:
        result.push(typeSymbol(statement, lineStarts));
        break;
      default:
        break;
    }
  }
  return result;
}
